REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [508]:
import pandas as pd

In [509]:
from zipfile import ZipFile

file_name = 'data.zip'

with ZipFile(file_name, 'r') as zip:
    zip.printdir()
    
    print('Extracting all files now...')
    zip.extractall()
    print('Done!')

File Name                                             Modified             Size
data/businesses.csv                            2017-10-05 18:06:50       660035
data/inspections.csv                           2017-10-05 18:06:50       505325
data/scoreCard.jpg                             2017-10-05 18:06:50        80875
Extracting all files now...
Done!


In [510]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [511]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [512]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [513]:
print(len(bus.index), 'records')

6315 records


#### Question 2b: How many unique business IDs are there?  

In [514]:
Unique_bIDS = len(pd.unique(bus['business_id']))
print(Unique_bIDS, 'unique business IDs')

6315 unique business IDs


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [517]:
bus['name'].value_counts()

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
                              ..
GOLDEN BURMA RESTAURANT        1
MUKA                           1
TUBA RESTAURANT                1
COVA HOTEL                     1
JOY'S PLACE                    1
Name: name, Length: 5692, dtype: int64

In [518]:
bus['name'].value_counts().sum()

6315

In [519]:
bus['name'].value_counts().head() #discovered entire dataset is in San Francisco

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

In [520]:
len(bus[bus.city == 'San Francisco']) #^^evidence same as number of records

6315

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [521]:
import numpy as np
bus['postal_code'].dtypes

dtype('O')

In [522]:
bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

#### Question 3b

What are the unique values of postal_code?

In [523]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [524]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [525]:
vb = bus[bus['postal_code'].isin(validZip)]
print(len(vb.index), 'businesses')

5999 businesses


In [526]:
vb

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
6310        NaN         NaN           NaN  
6311        NaN         NaN  +14154945048  
6312        NaN         NaN  +14158606036  
6313        NaN         NaN  +14153331878  
6314        NaN         NaN  +14158881393  

[5999 rows x 9 columns]

In [530]:
g = vb[['longitude', 'postal_code']].groupby('postal_code') #same as using the not_null_lon
g.count() 

longitude
postal_code           
94102              262
94103              306
94104               74
94105              126
94107              204
94108              142
94109              237
94110              334
94111              167
94112              131
94114              122
94115              162
94116               58
94117              130
94118              143
94121              106
94122              147
94123              116
94124               76
94127               43
94131               34
94132               69
94133              282
94134               45

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [531]:
null_lon = vb[pd.isnull(vb['longitude'])]
null_lon['business_id'].value_counts()

77826    1
82649    1
78545    1
84690    1
33491    1
        ..
5488     1
77169    1
5490     1
85363    1
82992    1
Name: business_id, Length: 2483, dtype: int64

In [532]:
null_lon

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

In [533]:
print ('2483 businesses missing longitudes')

2483 businesses missing longitudes


In [535]:
not_null_lon = vb[pd.isnull(vb['longitude']) == False]
not_null_lon['business_id'].value_counts()

826      1
9714     1
3543     1
7641     1
3547     1
        ..
68356    1
773      1
37639    1
35592    1
2047     1
Name: business_id, Length: 3516, dtype: int64

In [536]:
not_null_lon

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
4025        71784                    URBAN BREAD CAFE   
4029        71823            SUBWAY SANDWICHES #24254   
4031        71841                  RAPHA RACHING, LLC   
4033        71909            SHANGHAI DUMPLING KING 2   
4035        71915              SAPPHIRE ASIAN CUISINE   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
4025                  3901 18TH ST   San Francisco    CA       94114   
4029               200 PINE ST #140  San Francisco    CA       94104   
4031               2198 FILBERT ST   San Francisco    CA       94123   
4033             696 MONTEREY BLVD   San Francisco    CA       94127   
4035             475 SACRAMENTO ST   San Francisco    CA       94111   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
4025  37.761113 -122.430789  +14155745682  
4029  37.792237 -122.400124  +14152386388  
4031  37.798078 -122.435444  +14152063087  
4033  37.731568 -122.450970  +14152382380  
4035  37.794063 -122.400726  +14153781469  

[3516 rows x 9 columns]

In [537]:
print('3516 business not missing longitudes')

3516 business not missing longitudes


#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [538]:
nn_count = not_null_lon[['business_id', 'postal_code']].groupby('postal_code') #used business IDs because can count those
nn_count = nn_count.count()                                                    #cannot count the NaNs in the null_lon
nn_count['business_id']
nn_count = nn_count.rename(columns={"business_id":"not_null_lon"})
nn_count

not_null_lon
postal_code              
94102                 262
94103                 306
94104                  74
94105                 126
94107                 204
94108                 142
94109                 237
94110                 334
94111                 167
94112                 131
94114                 122
94115                 162
94116                  58
94117                 130
94118                 143
94121                 106
94122                 147
94123                 116
94124                  76
94127                  43
94131                  34
94132                  69
94133                 282
94134                  45

In [539]:
n_count = null_lon[['business_id', 'postal_code']].groupby('postal_code') #refer to commn
n_count = n_count.count()
n_count['business_id']
n_count = n_count.rename(columns={"business_id":"null_lon"})
n_count

null_lon
postal_code          
94102             196
94103             252
94104              59
94105             100
94107             247
94108              82
94109             146
94110             251
94111             115
94112              59
94114             108
94115              76
94116              35
94117              67
94118             100
94121              55
94122             123
94123              53
94124              88
94127              28
94131              14
94132              65
94133             135
94134              29

In [540]:
tot_df = pd.merge(n_count, nn_count, on = ['postal_code'])
tot_df

null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134              29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [546]:
tot_df['null/tot'] = tot_df['null_lon'] / (tot_df['null_lon'] + tot_df['not_null_lon'])
tot_df_sort = tot_df.sort_values('null/tot',
                                     ascending=False)
tot_df_sort.head()

null_lon  not_null_lon  null/tot
postal_code                                  
94107             247           204  0.547672
94124              88            76  0.536585
94132              65            69  0.485075
94114             108           122  0.469565
94122             123           147  0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [568]:
ins

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15430 rows x 4 columns]

In [569]:
a = len(ins.index)
a

15430

In [570]:
ins['business_id'].unique()

Unique_bIDS = len(pd.unique(ins['business_id']))
print(Unique_bIDS, 'unique business IDs')

5730 unique business IDs


In [611]:
print("There's more than one inspection per business because",  a, "rows/inspections do not match the number of unique businesses:", Unique_bIDS)

There's more than one inspection per business because 15430 rows/inspections do not match the number of unique businesses: 5730


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [572]:
ins['type'].unique()

array(['routine', 'complaint'], dtype=object)

In [573]:
ins.type.value_counts()

routine      15429
complaint        1
Name: type, dtype: int64

In [574]:
ins.dtypes

business_id     int64
score           int64
date            int64
type           object
dtype: object

In [575]:
ins2 = ins.copy()

In [576]:
check = ins2[['business_id', 'type']].groupby('type')
check.count()

business_id
type                  
complaint            1
routine          15429

In [577]:
ins2.dtypes

business_id     int64
score           int64
date            int64
type           object
dtype: object

In [578]:
ins2_plus10 = ins2.groupby('type').filter(lambda x: len(x) > 10) #filters the count of types greater than 10
ins2_plus10.type.value_counts() 

routine    15429
Name: type, dtype: int64

In [579]:
ins2_equal1 = ins2.groupby('type').filter(lambda x: len(x) == 1) # passes check to make sure this works for complaint
ins2_equal1.type.value_counts()

complaint    1
Name: type, dtype: int64

In [580]:
ins2 = ins2_plus10 #makes for easier variable name for future code

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [581]:
from datetime import datetime
year = []
for x in ins2['date'].astype(str):
    year.append(datetime.strptime(x,'%Y%m%d').strftime('%Y'))

ins2['year'] = year
ins2.dtypes

business_id     int64
score           int64
date            int64
type           object
year           object
dtype: object

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [582]:
ins2.year.value_counts()


2016    5839
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [586]:
ins3_plus50 = ins2.groupby('year').filter(lambda x: len(x) > 50) #filters the year count of types greater than 50
ins3_plus50.year.value_counts()  
ins3 = ins3_plus50
ins3['year'] = ins3['year'].astype(int)
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15391 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [587]:
ins3.dtypes

business_id     int64
score           int64
date            int64
type           object
year            int32
dtype: object

In [588]:
ins4 = ins3.loc[ins3['year'] == 2016]
ins4

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [612]:
#vb.head(20)

In [607]:
bus_ins4 = pd.merge(vb,ins4, how = 'left')
bus_ins4

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4              31      NORMAN'S ICE CREAM AND FREEZES   
...           ...                                 ...   
7265        90234                ACCESS SFUSD THE ARC   
7266        90236                      PHO DONG HUONG   
7267        90237                    WIN'S RESTAURANT   
7268        90238                    USA POWER MARKET   
7269        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
4              2801 LEAVENWORTH ST   San Francisco    CA       94133   
...                             ...            ...   ...         ...   
7265                1520 HOWARD ST   San Francisco    CA       94103   
7266             667 MONTEREY BLVD   San Francisco    CA       94127   
7267               3040 TARAVAL ST   San Francisco    CA       94116   
7268                1524 OCEAN AVE   San Francisco    CA       94112   
7269               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  score        date     type    year  
0     37.791116 -122.403816  +14154217044   82.0  20160503.0  routine  2016.0  
1     37.786848 -122.421547  +14157763262   94.0  20160513.0  routine  2016.0  
2     37.792888 -122.403135  +14156779494   98.0  20161005.0  routine  2016.0  
3     37.792888 -122.403135  +14156779494   96.0  20160311.0  routine  2016.0  
4     37.807155 -122.419004           NaN   84.0  20160816.0  routine  2016.0  
...         ...         ...           ...    ...         ...      ...     ...  
7265        NaN         NaN           NaN    NaN         NaN      NaN     NaN  
7266        NaN         NaN  +14154945048    NaN         NaN      NaN     NaN  
7267        NaN         NaN  +14158606036    NaN         NaN      NaN     NaN  
7268        NaN         NaN  +14153331878    NaN         NaN      NaN     NaN  
7269        NaN         NaN  +14158881393    NaN         NaN      NaN     NaN  

[7270 rows x 13 columns]

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [608]:
keep = ['name', 'address', 'score']
bus_ins4_low = bus_ins4.sort_values('score', ascending=True)
Final = bus_ins4_low[keep]
Final.head(20)

name              address  score
4755      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
4736                   UNCLE CAFE       65 WAVERLY PL    55.0
6042                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
7020                    POKI TIME     2101 LOMBARD ST    55.0
5448                   GOLDEN WOK   295 B ORIZABA AVE    56.0
2968             HING WANG BAKERY        339 JUDAH ST    56.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
5861    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
4107      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
4766  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
5214                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6386    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
4130                        ITHAI         720 POST ST    59.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
3498                  JADE GARDEN      3608 BALBOA ST    59.0
6657         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
4864                 WOLFES LUNCH        1220 16TH ST    60.0
5938          VALLARTA'S TACO BAR     2022 MISSION ST    60.0
4713             HONG KONG LOUNGE     5322 GEARY BLVD    60.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!